### ISSUE TITLE
Investigate CouchDB 3.x with APS/XPS frontend

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#11068]
**Impact of the new feature**
Most of the WM services

**Is your feature request related to a problem? Please describe.**
Somehow. Since CouchDB 2.x, all access to couchdb databases have to be authenticated (user/pass, token, proxy server, etc).
This might become challenging with the current setup and Apache CMSWEB frontends.

**Describe the solution you'd like**
As discussed with Valentin, we are investigating CouchDB 3.x deployment in the cmsweb dev clusters, using the APS (or XPS?) frontend as a reverse proxy to the CouchDB service.

What should be expected from this ticket is:
* the required yaml/configuration changes to deploy vanilla CouchDB 3.x in the cmsweb dev clusters
* verify functionality of the service (is it running properly? can we make http calls to it?)
* verify whether replication works and with which authentication method.

We should document all of this as well.

**Describe alternatives you've considered**
Alternative is using RPM-based CouchDB - PR available at the cmsdist repo - with our standard Apache frontends.

**Additional context**
Sub-task of https://github.com/dmwm/WMCore/issues/8853

**Comment by vkuznet** [comment]
I'm posing my findings from google doc over here to keep full description of APS+CouchDB in github issue.

I used cmsweb-test1 and cmsweb-test5 k8s clusters where I installed the following components:
- CouchDB image deployed via [couchdb.yaml](https://gist.github.com/vkuznet/a737f965e8e0b03aefefca6db1f81b08) k8s manifest file into couchdb namespace
- CouchDB configuration file local.ini provided by Alan
- APS and XPS deployed as daemon sets using the following redirect rule
```
       {
        "path":"/couchdb.*",
        "service_url":"http://login:password@couchdb.couchdb.svc.cluster.local:5984",
            "old_path": "/couchdb",
            "new_path": ""
    },
```
Here the login and password are CouchDB administrative credentials. The `old_path` and `new_path` are used by APS/XPS as substitution from incoming client’s HTTP requests to HTTP redirect requests. Here is an example of HTTP request flow:
```
(client) https://cmsweb-test1.cern.ch/couchdb/testdb ->
(APS) http://login:password@couchdb.couchdb.svc.cluster.local:5984/testdb
```
Please notice here, the incoming request has `/couchdb/testdb` end-point, while redirect requests only `/testdb`.

Using this configuration I was able to test the following scenarios:
- Access to couchdb via https://cmsweb-testx.cern.ch/couchdb and https://cmsweb-testx.cern.ch:8443/couchdb URLs
  - The former was used for access based on OAuth tokens, while later for access using X509 certificates
- Create database into CouchDB via HTTP PUT request
- Create new documents in CouchDB via HTTP PUT requests
- Perform database replication using OAuth token approach, see all steps listed in [github/issue#8853](https://github.com/dmwm/WMCore/issues/8853)

Here are the steps associated with access and replication:
Access CouchDB
```
# using token
token=`oidc-token vocms0181-test` # obtain token via oidc-token tool
curl -H "Authorization: bearer $token" https://cmsweb-test1.cern.ch/couchdb/testdb/_all_docs
    
# access via X509 cert
curl -L -k --key ~/.globus/userkey.pem --cert ~/.globus/usercert.pem \
          https://cmsweb-test1.cern.ch/couchdb/testdb/_all_docs 
```
Create DB
 ```
# using token
curl -k -X PUT -H "Authorization: bearer $token" https://cmsweb-test1.cern.ch/couchdb/testdb

# using X509
curl -L -k --key ~/.globus/userkey.pem --cert ~/.globus/usercert.pem -X PUT https://cmsweb-test1.cern.ch/couchdb/testdb
```
Create new document
```
curl -k -X PUT -H "Authorization: bearer $token" https://cmsweb-test1.cern.ch/couchdb/testdb/"001" -d '{"foo":1}'
```
Create replication doc
```
cat > replication.json EOF
{ "_id": "replication_job",
    "source": {
       "url": "https://cmsweb-test1.cern.ch/couchdb/testdb",
       "headers": { "Authorization": "Bearer xxxx" }
    },
    "target": {
       "url": "https://cmsweb-test5.cern.ch/couchdb/testdb",
       "headers": { "Authorization": "Bearer xxxx" }
    },
    "create_target": true,
    "continuous": false
}
EOF
```
Start replication
```
curl -k -X POST -H "Authorization: bearer $token" \
       -H "Content-Type: application/json" \
       -d@./replication.json https://cmsweb-test1.cern.ch/couchdb/_replicate
```
The replication via X509 mechanism using XPS on port 8443 does not work. And, as I realised, it will not work by design. Here is the explanation:
- If we use X509 authentication provided by reverse-proxy service (e.g. XPS or apache) and stock CouchDB image the access to CouchDB is provided by XPS.
- The XPS inspects provided X509 certificates from clients and sets custom CMS HTTP headers which is passed to back-end services, in our case to stock CouchDB
   - But stock CouchDB does not care about our CMS HTTP headers, unless we’ll set them up explicitly in the replication document (CouchDB supports the “headers” attribute in its replication document).
- In order to perform authentication the XPS (or apache) performs the following steps:
   - It extracts CMS specific headers, like Cms-Auth-Cert, Cms-Authn-DN
   - It compares DN with list of available DNs from CRIC
   - It calculates hash using CMS headers and compare it with the one used by XPS (apache) hmac
- For HTTP accesses to CouchDB X509 mechanism works since incoming HTTP request which client provides satisfies all above steps, e.g. we can access CouchDB using the following command:
```
curl -L -k --key $X509_USER_PROXY --cert $X509_USER_PROXY \
             https://cmsweb-test1.cern.ch:8443/couchdb/testdb
```
- For CouchDB replication this process does not work since the initial replication request reaches the CouchDB where the CouchDB process spawns its own HTTP process between two hosts (source and target). But since the original replication request contains cmsweb hostname, the HTTP requests between two couches can’t be fulfilled due to the fact that CouchDB which initiated this request does not have the user's X509 certificate which is required during TLS handshake between client and the host:
    - The replication document (request) contains source and target URLs which contains our FEs hostnames, e.g. “source”:”https://cmsweb-test1.cern.ch:8443/couchdb/testbed” and “target”:”https://cmsweb-test5.cern.ch:8443/couchdb/testdb” 
   - In order to access either source or target URLs client should provide X509 certificates, otherwise this request will be rejected by XPS (apache).
   - Since the replication requests will be initiated by CouchDB instance and we use stock couch image it does not have our client’s certificates
   - In order to make replication work the client (in this case CouchDB process itself) must provide an X509 certificate to the server, and this requires changes to CouchDB itself where we need to pass client’s certificate.

But CouchDB replication using token based authentication works just fine. Here are the reasons:
- The replication document (request) contains not only source and target URLs, but also HTTP headers with specific token
```
       { "_id": "replication_job",
         "source": {
           "url": "https://cmsweb-test1.cern.ch/couch/testdb",
           "headers": { "Authorization": "Bearer xxxx" }
         },
         "target": {
           "url": "https://cmsweb-test5.cern.ch/couch/testdb",
           "headers": { "Authorization": "Bearer xxxx" }
         },
         "create_target": true,
         "continuous": false
       }
```
- The headers are used by the CouchDB process when a replication request is initiated. Therefore, the CouchDB HTTP request contains a proper HTTP header which will be accepted by XPS since it contains a valid token.
- This process will continue to run correctly until the authorization token is valid, e.g. if we have a continuous replication process we must ensure that the token should be valid the entire time of the replication process.

**Comment by amaltaro** [comment]
When this issue was created, we were considering some tweaks to the frontend code such that it would ease the process to upgrade to the latest CouchDB version. In the end we managed to get it working without any frontend changes, however, I think this is still to be considered such that we can migrate CMSWEB frontends in the coming months/quarters. Moving to Q3

**Comment by vkuznet** [comment]
@amaltaro , since this ticket among High-Priority I would like to clarify how we should move with this. We can easily perform switch of FE (from apache to aps) on testbed but before we do that we need input from you about the following topics:
- how to setup replication, as I listed in this ticket it will require tokens which has expiration time. As such I doubt continuous replication is possible and I need clear instructions from you if it will or will not cause any disruptions on WMCore services and how it should be organized
- If we perform the switch the CLI tools (accessing port 8443) should start using tokens and I do not know if WMCore tools and services are ready for that
- it is still unclear if web UI is fully tested

Meanwhile, I'll start review of redirect rules and coordinate with CMSWEB team the required changes with APS/XPS configs if necessary.

**Comment by amaltaro** [comment]
@vkuznet as we just discussed, we should first try to have the full WMCore **dev** environment operational, hence working on this CouchDB issue first (thanks for taking it over!):
https://github.com/dmwm/WMCore/issues/11086

Once we have this setup, we can then start testing things in pre pre-production to have a better confidence degree.
However, I do think that we will have to migrate to APS without tokens enabled. I think having some real production experience with a new frontend application is important before we can migrate both frontends and authentication mechanism.

**Comment by vkuznet** [comment]
@amaltaro , there is no way to decouple proxy server and auth procedure as it is one code flow (first auth and then proxy). As such we will need to test APS (which is token based authentication) with tokens. But the proxy part is standard library of Go and as such we can test it using x509 authentication (XPS). To make it clear we have the following request flow in APS/XPS:
```
client -> (token or x509) -> auth part (APS/XPS) -> Go proxy
         |--------------------------------------------------|
                            auth-proxy-server
```

**Comment by amaltaro** [comment]
> In other words, we can upgrade FE to use XPS on both ports (443 and 8443) which will handle x509 and provide proxy server. Then, once we get experience we'll switch port 443 to use APS which will use token authentication.

Yes, this is more or less what I have in mind. We need to start with a setup that is known for being functional. Then we start changing it towards the end goal.

BTW, for the record, here are the required changes for the testX cluster frontends:
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/commit/a5957c4a20fd4984afdcf4328280d663a5958869

note that the cmsweb-testX url still has to be adapted (the HTTP team is working on fixing this automatically).
We also have the redirect rules in this file, in case we have problems with services at a given endpoint, we know where to look at.

**Comment by vkuznet** [comment]
Alan, the main point is that you should start adopting end-to-end deployment procedure. We have different groups and different schedules. As such, CMSWEB group provides all the tools required to do what each group need. CMSWEB group goal is to move towards token based authentication and need other services to perform validation. Therefore, if you will take over your clusters you can
- change APS with XPS if you need to or rollback to apache FEs
- you will not interfere with other groups and schedules
- you can work on your schedule towards your milestones

As I wrote, we can easily swap on FE apache with either APS/XPS, but I doubt we will do it for all testX clusters. Please consider this and if for dmwm clusters you need this procedure we can either give you rights and show you how to do it, or you can request the swap from CMSWEB group of specific FE on specific testX clusters. All other clusters will continue to use APS for standard port (443) and XPS for machine port 8443.

**Comment by vkuznet** [comment]
@amaltaro , @goughes, @todor-ivanov I performed testing of CouchDB replication using APS/XPS and put detailed description to this [gist](https://gist.github.com/vkuznet/e7f31ceee893df10e29e2c21eb92bd44). Please read it and feel free to replicate all steps in your setup (clusters).

Here is executive summary:
- replication via token based authentication (via APS, port 443) works
- replication via x509 based authentication (via XPS, port 8443) does not work since CouchDB requires code to read and handle x509 certificate in replication process
- data injection via x509 based authentication (via XPS, port 8443) works
- data injection via token based authentication (via APS, port 554) works but requires Referer: xxx HTTP header which we should add to APS.

**Comment by amaltaro** [comment]
@vkuznet hi Valentin, thanks for performing these tests. I have a few comments on the executive summary:

> replication via x509 based authentication (via XPS, port 8443) does not work since CouchDB requires code to read and handle x509 certificate in replication process

this cannot be the reason! We run CouchDB replication between WMAgents and central CouchDB using certificates. It's just a matter of having the correct `[ssl]` and `[replicator]` sections in the node starting replications (even though I assume you have it, otherwise I don't know how you would test replication...)

> data injection via token based authentication (via APS, port 554) works but requires Referer: xxx HTTP header which we should add to APS.

this `Referer` header seems to carry the exact url being requested by the client. Is it supposed to be provided by the client or APS? It is strange thoough that token based authentication would require this header while certificated-based one does not.

**Comment by vkuznet** [comment]
The local.ini in test branch on gitlab does not have any ssl/replicator sections and this is what I used. Moreover, as I wrote in https://github.com/dmwm/WMCore/issues/10452 I do not know if sock couchdb image supports x509. If it does we need to test it and adjust local.ini accordingly. I'm testing whatever you put in test branch and so far it does not contains ssl/replicator sections.

Regarding `Referer`, I need to fix APS/XPS code for this header, but nothing strange since x509 authentication didn't authenticate (yet) and that's why the header issue was not present (it comes after authentication).

**Comment by vkuznet** [comment]
Here is corresponding commit to `local.ini` in test branch https://gitlab.cern.ch/cmsweb-k8s/services_config/-/commit/33020595ee4695a8243f0043954d22fcb62efe5a which is required to make x509 replication works in testX clusters

**Comment by vkuznet** [comment]
Meanwhile, I fixed/tested APS/XPS/SPS to setup Referrer, and requested to update them in k8s clusters via this ticket: https://its.cern.ch/jira/browse/CMSKUBERNETES-184 Once we'll deploy new version of APS/XPS/SPS they will have everything dev-team need to do their work with token and x509 authentication, including CouchDB replication (if it will be ever required).

**Comment by vkuznet** [comment]
@amaltaro From my point of view everything now is in working state, i.e. CouchDB is deployed to testX clusters, and its replication works both using x509 and token based authentication methods. I provided instructions, I requested APS/XPS/SPS changes to put in place and I request that testX clusters should have all namespaces and proxy/tokens in place by default on testX clusters. I also updated `local.ini` in test branch that we will have an example how to do this, and I tested it with test10/test5 clusters having local CouchDB and APS/XPS frontends. Do we need anything else in this ticket?
